### PR TITLE
177261 implement updated sap action rules

### PR DESCRIPTION
**Pull request by pravinvarade10801**
Implement updated SAP action rules
Added unit test cases

### CODE REVIEW

File: tests/UKHO.ERPFacade.API.FunctionalTests/appsettings.json
Comment (Vishal14583) [inline tests/UKHO.ERPFacade.API.FunctionalTests/appsettings.json:1]: Remove unused configuration of XMLActionList

File: src/UKHO.ERPFacade.Common/Logging/EventIds.cs [hunk 1]
```diff
@@ -704,15 +404,39 @@ public enum EventIds
         MessageDequeueCount = 940140,
 
         /// <summary>
-        /// 940141 - Invalid UkhoWeekNumber field in enccontentpublished event from EES.
+        /// 940142 - Exception occurred while decrypting the permit string.
```
Comment (Vishal14583) [inline src/UKHO.ERPFacade.Common/Logging/EventIds.cs:407]: Please correct the event id order.. Few event sequence numbers are missing.

File: tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/2NewCells.JSON [hunk 2]
```diff
@@ -0,0 +1,227 @@
+{
+  "specversion": "1.0",
+  "type": "uk.gov.ukho.encpublishing.enccontentpublished.v2.2",
+  "source": "https://encpublishing.ukho.gov.uk",
+  "id": "e7bd473c-a573-4905-9541-77218df10b1b",
+  "time": "2024-09-13T10:22:26.1824904+00:00",
+  "subject": "US5AK2AX,US5AK9DI",
+  "datacontenttype": "application/json",
+  "data": {
+    "correlationId": "e4c9d333-b133-4133-a0aa-08dccb35b407",
```
[the remark this answers is not in the document]
Comment (rushikeshkhopatkar) [reply tests/UKHO.ERPFacade.API.FunctionalTests/ERPFacadePayloadTestData/2NewCells.JSON]: Added prefix.

File: tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs [hunk 3]
```diff
@@ -1,1113 +1,117 @@
...
-                    //xmlAttributes[1] is skipped as already checked
-                    if (!item.PRODUCT.Equals("ENC CELL"))
-                        attrNotMatched.Add(nameof(item.PRODUCT));
-                    if (!item.PRODTYPE.Equals(product.ProductType[4..]))
-                        attrNotMatched.Add(nameof(item.PRODTYPE));
-                    VerifyAdditionalXmlTags(item, correctionTag);
-                    //Checking blanks
-                    List<string> blankFieldNames = new() { "PRODUCTNAME", "CANCELLED", "AGENCY", "PROVIDER", "ENCSIZE", "TITLE", "EDITIONNO", "UPDATENO", "UNITTYPE", "ACTIVEKEY", "NEXTKEY" };
-                    VerifyBlankFields(item, blankFieldNames);
+            var generatedItems = generatedXml.Descendants("Item").ToList();
```
[the remark this answers is not in the document]
Comment (rushikeshkhopatkar) [reply tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs]: Updated as suggested.

File: tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs [hunk 4]
```diff
@@ -1,1113 +1,117 @@
...
-                    Console.WriteLine("Not matching attributes are:");
-                    foreach (string attribute in attrNotMatched)
-                    { Console.WriteLine(attribute); }
-                    return false;
-                }
-            }
-            Console.WriteLine("JSON doesn't have corresponding Unit of Sale.");
-            return false;
-        }
+                var item1 = generatedItems[i];
```
[the remark this answers is not in the document]
Comment (rushikeshkhopatkar) [reply tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs]: Updated 'item1' to 'generatedAction' and 'item2' to 'expectedAction'

File: tests/UKHO.ERPFacade.API.FunctionalTests/UKHO.ERPFacade.API.FunctionalTests.csproj [hunk 5]
```diff
@@ -23,7 +23,8 @@
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="FluentAssertions" Version="6.12.1" />
+    <PackageReference Include="Azure.Security.KeyVault.Secrets" Version="4.6.0" />
```
[the remark this answers is not in the document]
Comment (rushikeshkhopatkar) [reply tests/UKHO.ERPFacade.API.FunctionalTests/UKHO.ERPFacade.API.FunctionalTests.csproj]: Removed.

File: tests/UKHO.ERPFacade.API.FunctionalTests/appsettings.json
Comment (rushikeshkhopatkar) [reply tests/UKHO.ERPFacade.API.FunctionalTests/appsettings.json:1]: Removed.

File: src/UKHO.ERPFacade.Common/Logging/EventIds.cs (hunk 1 above)
Comment (pravinvarade10801) [reply src/UKHO.ERPFacade.Common/Logging/EventIds.cs]: Event Id sequence updated.

File: src/UKHO.ERPFacade.API/Helpers/EncContentSapMessageBuilder.cs [hunk 6]
```diff
@@ -66,478 +73,406 @@ IPermitDecryption permitDecryption
...
             string sapXmlTemplatePath = Path.Combine(Environment.CurrentDirectory, SapXmlPath);
 
-            //Check whether template file exists or not
+            // Check if SAP XML payload template exists
             if (!_fileSystemHelper.IsFileExists(sapXmlTemplatePath))
             {
-                _logger.LogError(EventIds.SapXmlTemplateNotFound.ToEventId(), "The SAP message xml template does not exist.");
-                throw new FileNotFoundException();
+                _logger.LogError(EventIds.SapXmlTemplateNotFound.ToEventId(), "The SAP xml payload template does not exist.");
+                throw new FileNotFoundException("The SAP xml payload template does not exist.");
```
[the remark this answers is not in the document]
Comment (pravinvarade10801) [reply src/UKHO.ERPFacade.API/Helpers/EncContentSapMessageBuilder.cs]: Done.

File: src/UKHO.ERPFacade.Common/Logging/EventIds.cs [hunk 7]
```diff
@@ -56,663 +56,387 @@ public enum EventIds
...
-        /// 940141 - Invalid UkhoWeekNumber field in enccontentpublished event from EES.
-        /// </summary>
-        InvalidUkhoWeekNumber = 940141,
-
-        /// <summary>
-        /// 940141 - Exception occured while decrypting the permit string.
-        /// </summary>
-        PermitDecryptionException = 940142
-
+        RequiredSectionNotFound = 940086
```
Comment (Vishal14583) [inline src/UKHO.ERPFacade.Common/Logging/EventIds.cs:439]:
```suggestion
        RequiredSectionNotFoundException = 940086
```

File: tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs [hunk 8]
```diff
@@ -1,1113 +1,114 @@
...
-                    if (!item.PRODUCT.Equals("ENC CELL"))
-                        attrNotMatched.Add(nameof(item.PRODUCT));
-                    if (!item.PRODTYPE.Equals(product.ProductType[4..]))
-                        attrNotMatched.Add(nameof(item.PRODTYPE));
-                    VerifyAdditionalXmlTags(item, correctionTag);
-                    //Checking blanks
-                    List<string> blankFieldNames = new() { "PRODUCTNAME", "CANCELLED", "AGENCY", "PROVIDER", "ENCSIZE", "TITLE", "EDITIONNO", "UPDATENO", "UNITTYPE", "ACTIVEKEY", "NEXTKEY" };
-                    VerifyBlankFields(item, blankFieldNames);
+            var generatedAttributes = generatedXml.Descendants("Item").ToList();
+            var expectedItems = expectedXml.Descendants("Item").ToList();
```
Comment (Vishal14583) [inline tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs:68]:
```suggestion
            var expectedAttributes = expectedXml.Descendants("Item").ToList();
```

File: tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs [hunk 9]
```diff
@@ -1,1113 +1,114 @@
...
-            return false;
-        }
-
-        private static bool VerifyCreateEncCell(string childCell, ZMAT_ACTIONITEMS item, string correctionTag, string permitState)
-        {
-            Console.WriteLine("Action#:" + actionCounter + ".Childcell:" + childCell);
-            foreach (Product product in JsonPayload.Data.Products)
-            {
-                if ((childCell == product.ProductName) && (product.Status.IsNewCell))
+                foreach (var element1 in generatedAction.Elements())
```
Comment (Vishal14583) [inline tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs:88]: Can we have meaningful variable name please

File: tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs [hunk 10]
```diff
@@ -1,1113 +1,114 @@
...
-                    Console.WriteLine("Not matching attributes are:");
-                    foreach (string attribute in attrNotMatched)
-                    { Console.WriteLine(attribute); }
-                    return false;
-                }
-            }
-            Console.WriteLine("JSON doesn't have corresponding product.");
-            return true;
-        }
+                    var element2 = expectedAction.Element(element1.Name);
```
Comment (Vishal14583) [inline tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs:90]: Meaningful name here as well

File: tests/UKHO.ERPFacade.API.FunctionalTests/FunctionalTests/WebhookScenarios.cs
Comment (Vishal14583) [inline tests/UKHO.ERPFacade.API.FunctionalTests/FunctionalTests/WebhookScenarios.cs:1]: Remove unused usings

**Review by Vishal14583 — CHANGES_REQUESTED** [review]
Few more review comments to resolve.

File: tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs (hunk 9 above)
Comment (pravinvarade10801) [reply tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs]: done

File: src/UKHO.ERPFacade.Common/Logging/EventIds.cs (hunk 7 above)
Comment (pravinvarade10801) [reply src/UKHO.ERPFacade.Common/Logging/EventIds.cs]: done

File: tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs (hunk 10 above)
Comment (pravinvarade10801) [reply tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs]: done

File: tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs (hunk 8 above)
Comment (pravinvarade10801) [reply tests/UKHO.ERPFacade.API.FunctionalTests/Helpers/SAPXmlHelper.cs]: done

**Review by Vishal14583 — APPROVED** [review]
Looks good to me.

**Review by rushikeshkhopatkar — APPROVED** [review]
Looks good

**Review by Vishal14583 — APPROVED** [review]

**Review by rushdynajath — APPROVED** [review]
Looks good

**Pull request merged**